### PR TITLE
KIBA-1533: Create a process to index tokens by attribute

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**
```
1. Every X minutes we send UPDATE_ALL_COLLECTION_TOKEN_ATTRIBUTES to token queue
2. On receiving UPDATE_ALL_COLLECTION_TOKEN_ATTRIBUTES:
   2.1 Get the current time
   2.2 Get the last time this update all process ran from db
   2.3 Get all tokens in certain collections that have been changed since 2.2
   2.4 For each token in 2.3:
       2.4.1 Add a UPDATE_COLLECTION_TOKEN_ATTRIBUTES to token queue
   2.5 Save 2.1 as the last time this all process ran in the db
3. On receiving UPDATE_COLLECTION_TOKEN_ATTRIBUTES:
   3.1 Get the token's attributes from the metadata table in db
   3.2 Get the token's currently saved attributes in attributes table
   3.3 For every attribute in 3.2 :
       3.3.1 Delete attribute from attributes table
   3.4 For every attribute in 3.1:
       3.4.1 Save attribute to attributes table
 
```

### CODE REVIEW

File: api/notd/api/api_v1.py [hunk 1]
```diff
@@ -88,6 +89,11 @@ async def update_activity_for_all_collections_deferred():
         await notdManager.update_activity_for_all_collections_deferred()
         return UpdateActivityForAllCollectionsDeferredResponse()
 
+    @router.post('/collections/update-attributes-deferred', response_model=UpdateAttributeForAllTokensDeferredResponse)
```
Comment (krishan711) [inline api/notd/api/api_v1.py:92]: this path is wrong. try to think of something better. if you cant i will but you try first

File: api/notd/api/endpoints_v1.py [hunk 2]
```diff
@@ -163,3 +163,9 @@ class ListCollectionTokenAirdropsRequest(BaseModel):
 
 class ListCollectionTokenAirdropsResponse(BaseModel):
     airdrops: List[ApiAirdrop]
+
+class UpdateAttributeForAllTokensDeferredRequest(BaseModel):
```
Comment (krishan711) [inline api/notd/api/endpoints_v1.py:167]: the plularisation in this is wrong. i think the name can be much better too. it def needs collection in it somwhere. please look at how other messages are named. they follow they same pattern everywhere and it starts with the api path.

File: api/notd/model.py [hunk 3]
```diff
@@ -251,3 +251,16 @@ class Airdrop:
     isClaimed: bool
     claimTokenKey: Token
     claimUrl: str
+
+@dataclasses.dataclass
+class RetrievedTokenAttribute:
+    registryAddress: str
+    tokenId: str
+    attributeName: Optional[str]
```
Comment (krishan711) [inline api/notd/model.py:259]: is this right? why do you update it in the next pr but not fix here?

File: api/notd/token_manager.py [hunk 4]
```diff
@@ -396,6 +407,28 @@ async def update_activity_for_all_collections(self) -> None:
         await self.tokenQueue.send_messages(messages=messages)
         await self.saver.update_latest_update(latestUpdateId=latestUpdate.latestUpdateId, date=startDate)
 
+    async def update_attribute_for_all_tokens(self) -> None:
+        #Get date of the last processed attribute
+        #Get all tokens to have been updated after this date
+        #send all the updatedTokens to the queue for processing
+        latestProcessedTokenAttribute: List[TokenAttribute] = await self.retriever.list_token_attributes(orders=[Order(fieldName=TokenAttributesTable.c.updatedDate.key, direction=Direction.DESCENDING)], limit=1)
```
Comment (krishan711) [inline api/notd/token_manager.py:414]: what is this? why not use our new table instead of this?

File: api/notd/token_manager.py [hunk 5]
```diff
@@ -396,6 +407,28 @@ async def update_activity_for_all_collections(self) -> None:
...
+        if len(latestProcessedTokenAttribute) > 0:
+            latestProcessedDate = latestProcessedTokenAttribute[0].updatedDate
+        else:
+            latestProcessedDate = date_util.datetime_from_now(weeks=-1)
+        logging.info(f'Finding changed tokens since {latestProcessedDate}')
+        updateTokenMetadataQuery = (
```
Comment (krishan711) [inline api/notd/token_manager.py:420]: this varaible name is terrible - are you updating something with this query?

File: api/application.py [hunk 6]
```diff
@@ -62,6 +63,9 @@
 collectionProcessor = CollectionProcessor(requester=requester, ethClient=ethClient, openseaApiKey=openseaApiKey, s3manager=s3manager, bucketName=os.environ['S3_BUCKET'])
 tokenOwnershipProcessor = TokenOwnershipProcessor(retriever=retriever)
 collectionActivityProcessor = CollectionActivityProcessor(retriever=retriever)
+tokenAttributeProcessor = TokenAttributeProcessor(retriever=retriever)
+tokenManager = TokenManager(saver=saver, retriever=retriever, tokenQueue=tokenQueue, collectionProcessor=collectionProcessor, tokenMetadataProcessor=tokenMetadataProcessor, tokenOwnershipProcessor=tokenOwnershipProcessor, collectionActivityProcessor=collectionActivityProcessor, tokenAttributeProcessor=tokenAttributeProcessor)
+revueApiKey = os.environ['REVUE_API_KEY']
```
Comment (krishan711) [inline api/application.py:68]: rm this line

File: api/notd/api/api_v1.py [hunk 7]
```diff
@@ -94,6 +95,11 @@ async def update_activity_for_all_collections_deferred():
         await notdManager.update_activity_for_all_collections_deferred()
         return UpdateActivityForAllCollectionsDeferredResponse()
 
+    @router.post('/collection/update-all-token-attributes-deferred', response_model=UpdateAllCollectionTokenAttributesDeferredResponse)
```
Comment (krishan711) [inline api/notd/api/api_v1.py:98]:
```suggestion
    @router.post('/collections/update-all-token-attributes-deferred', response_model=UpdateAllCollectionTokenAttributesDeferredResponse)
```
```suggestion
    @router.post('/collection/update-all-token-attributes-deferred', response_model=UpdateAllCollectionTokenAttributesDeferredResponse)
```

File: api/notd/api/api_v1.py [hunk 8]
```diff
@@ -94,6 +95,11 @@ async def update_activity_for_all_collections_deferred():
...
+    async def update_all_collection_token_attributes_deferred():
+        await notdManager.update_all_collection_token_attributes_deferred()
```
Comment (krishan711) [inline api/notd/api/api_v1.py:100]: this should be update_token_attributes_for_all_collections (see the listings and activity example)

File: api/notd/manager.py [hunk 9]
```diff
@@ -347,15 +347,24 @@ async def update_collection_tokens(self, address: str, shouldForce: bool = False
...
+    async def update_all_collection_token_attributes(self) -> None:
+        await self.tokenManager.update_all_collection_token_attributes()
+
     async def update_activity_for_collection_deferred(self, registryAddress: str, startDate: datetime.datetime) -> None:
         await self.tokenManager.update_activity_for_collection_deferred(registryAddress=registryAddress, startDate=startDate)
 
     async def update_activity_for_collection(self, address: str, startDate: datetime.datetime) -> None:
         await self.tokenManager.update_activity_for_collection(address=address, startDate=startDate)
 
+    async def update_collection_token_attribute(self, registryAddress: str, tokenId: str) -> None:
```
Comment (krishan711) [inline api/notd/manager.py:365]: this needs to be plural, you are updating all the attributes

File: api/notd/messages.py [hunk 10]
```diff
@@ -52,6 +52,13 @@ class UpdateActivityForCollectionMessageContent(MessageContent):
     address: str
     startDate: datetime.datetime
 
+class UpdateAllCollectionTokenAttributesMessageContent(MessageContent):
+    _COMMAND = 'UPDATE_ALL_COLLECTION_TOKEN_ATTRIBUTES'
```
Comment (krishan711) [inline api/notd/messages.py:56]: UPDATE_TOKEN_ATTRIBUTES_FOR_ALL_COLLECTIONS

File: dbschema/1.tables.sql [hunk 11]
```diff
@@ -209,12 +209,25 @@ CREATE INDEX tbl_latest_updates_date ON tbl_latest_updates (date);
 CREATE INDEX tbl_latest_updates_key ON tbl_latest_updates (key);
 CREATE INDEX tbl_latest_updates_name ON tbl_latest_updates (name);
 
-CREATE TABLE tbl_latest_token_listings (
+
```
Comment (krishan711) [inline dbschema/1.tables.sql:212]: move your table below it will remove this error

File: dbschema/1.tables.sql [hunk 12]
```diff
@@ -209,12 +209,25 @@ CREATE INDEX tbl_latest_updates_date ON tbl_latest_updates (date);
...
+);
+CREATE UNIQUE INDEX tbl_token_attributes_registry_address_token_id_attribute_name on tbl_token_attributes (registry_address, token_id, attribute_name);
+CREATE INDEX tbl_token_attributes_created_date ON tbl_token_attributes (created_date);
+CREATE INDEX tbl_token_attributes_updated_date ON tbl_token_attributes (updated_date);
+CREATE INDEX tbl_token_attributes_registry_address ON tbl_token_attributes (registry_address);
+CREATE INDEX tbl_token_attributes_token_id ON tbl_token_attributes (token_id);
+CREATE INDEX tbl_token_attributes_attribute_name ON tbl_token_attributes (attribute_name);
+CREATE INDEX tbl_token_attributes_attribute_value ON tbl_token_attributes (attribute_value);
+
+CREATE TABLE tbl_latest_token_listings (
```
Comment (krishan711) [inline dbschema/1.tables.sql:230]: you need to add lines to 2.users.sql (see listings pr)

File: api/notd/store/schema_conversions.py [hunk 13]
```diff
@@ -168,6 +170,16 @@ def latest_update_from_row(row: Mapping) -> LatestUpdate:
         date=row[LatestUpdatesTable.c.date],
     )
 
+def token_attributes_from_row(row: Mapping) -> TokenAttribute:
```
Comment (krishan711) [inline api/notd/store/schema_conversions.py:173]: two lines between these

File: api/notd/store/schema.py [hunk 14]
```diff
@@ -154,15 +154,24 @@
...
+    sqlalchemy.Column(key='attributeValue', name='attribute_value', type_=sqlalchemy.Text, nullable=True),
+)
 
 LatestTokenListingsTable = sqlalchemy.Table(
     'tbl_latest_token_listings',
     metadata,
     sqlalchemy.Column(key='latestTokenListingId', name='id', type_=sqlalchemy.Integer, autoincrement=True, primary_key=True, nullable=False),
     sqlalchemy.Column(key='createdDate', name='created_date', type_=sqlalchemy.DateTime, nullable=False),
     sqlalchemy.Column(key='updatedDate', name='updated_date', type_=sqlalchemy.DateTime, nullable=False),
-    sqlalchemy.Column(key='registryAddress', name='registry_address', type_=sqlalchemy.Text, nullable=False),
```
Comment (krishan711) [inline api/notd/store/schema.py:164]: ????

File: api/notd/token_manager.py [hunk 15]
```diff
@@ -410,6 +420,24 @@ async def update_activity_for_all_collections(self) -> None:
...
+        logging.info(f'Finding changed tokens since {latestProcessedDate}')
+        recentMetadataTokenQuery = (
+            TokenMetadatasTable.select()
+            .with_only_columns([TokenMetadatasTable.c.registryAddress, TokenMetadatasTable.c.tokenId])
+            .where(TokenMetadatasTable.c.registryAddress == SPRITE_CLUB_REGISTRY_ADDRESS)
+            .where(TokenMetadatasTable.c.updatedDate >= latestProcessedDate)
+        )
+        recentMetadataTokenQueryResult = await self.retriever.database.execute(query=recentMetadataTokenQuery)
+        updatedTokenMetadatas = set(recentMetadataTokenQueryResult)
+        logging.info(f'Scheduling processing for {len(updatedTokenMetadatas)} updatedTokenAttributes')
```
Comment (krishan711) [inline api/notd/token_manager.py:436]: wrong message

File: api/notd/token_manager.py [hunk 16]
```diff
@@ -410,6 +420,24 @@ async def update_activity_for_all_collections(self) -> None:
         await self.tokenQueue.send_messages(messages=messages)
         await self.saver.update_latest_update(latestUpdateId=latestUpdate.latestUpdateId, date=startDate)
 
+    async def update_all_collection_token_attributes(self) -> None:
+        startDate = date_util.datetime_from_now()
+        latestUpdate = await self.retriever.get_latest_update_by_key_name(key='token_attributes')
+        latestProcessedDate = latestUpdate.date
+        logging.info(f'Finding changed tokens since {latestProcessedDate}')
+        recentMetadataTokenQuery = (
```
Comment (krishan711) [inline api/notd/token_manager.py:428]: updatedTokenMetadatasQuery

File: api/notd/api/api_v1.py [hunk 17]
```diff
@@ -94,6 +95,11 @@ async def update_activity_for_all_collections_deferred():
         await notdManager.update_activity_for_all_collections_deferred()
         return UpdateActivityForAllCollectionsDeferredResponse()
 
+    @router.post('/collections/update-all-token-attributes-deferred', response_model=UpdateTokenAttributesForAllCollectionsDeferredResponse)
```
Comment (krishan711) [inline api/notd/api/api_v1.py:98]:
```suggestion
    @router.post('/collections/update-token-attributes-deferred', response_model=UpdateTokenAttributesForAllCollectionsDeferredResponse)
```

File: api/notd/manager.py [hunk 18]
```diff
@@ -356,6 +356,18 @@ async def update_activity_for_collection_deferred(self, registryAddress: str, st
     async def update_activity_for_collection(self, address: str, startDate: datetime.datetime) -> None:
         await self.tokenManager.update_activity_for_collection(address=address, startDate=startDate)
 
+    async def update_token_attributes_for_all_collections_deferred(self) -> None:
+        await self.tokenManager.update_token_attributes_for_all_collections_deferred()
+
+    async def update_token_attributes_for_all_collection(self) -> None:
+        await self.tokenManager.update_token_attributes_for_all_collection()
+
+    async def update_token_attributes_for_collection_deferred(self, registryAddress: str, tokenId: str) ->None:
```
Comment (krishan711) [inline api/notd/manager.py:365]:
```suggestion
    async def update_token_attributes_for_collection_token_deferred(self, registryAddress: str, tokenId: str) ->None:
```

File: api/notd/messages.py [hunk 19]
```diff
@@ -52,6 +52,13 @@ class UpdateActivityForCollectionMessageContent(MessageContent):
     address: str
     startDate: datetime.datetime
 
+class UpdateTokenAttributesForAllCollectionsMessageContent(MessageContent):
+    _COMMAND = 'UPDATE_TOKEN_ATTRIBUTES_FOR_ALL_COLLECTIONS'
+
+class UpdateTokenAttributesForCollectionMessageContent(MessageContent):
```
Comment (krishan711) [inline api/notd/messages.py:58]:
```suggestion
class UpdateCollectionTokenAttributesMessageContent(MessageContent):
```

File: api/notd/messages.py [hunk 20]
```diff
@@ -52,6 +52,13 @@ class UpdateActivityForCollectionMessageContent(MessageContent):
...
+    _COMMAND = 'UPDATE_TOKEN_ATTRIBUTES_FOR_COLLECTION'
```
Comment (krishan711) [inline api/notd/messages.py:59]:
```suggestion
    _COMMAND = 'UPDATE_COLLECTION_TOKEN_ATTRIBUTES'
```

File: api/notd/store/retriever.py [hunk 21]
```diff
@@ -219,7 +222,19 @@ async def list_latest_updates(self, fieldFilters: Optional[Sequence[FieldFilter]
         latestUpdates = [latest_update_from_row(row) for row in result]
         return latestUpdates
 
-    async def get_latest_update_by_key_name(self, key: str, name: Optional[str] = None, connection: Optional[DatabaseConnection] = None) -> LatestUpdate:  # pylint: disable=invalid-name
+    async def list_token_attributes(self, fieldFilters: Optional[Sequence[FieldFilter]] = None, orders: Optional[Sequence[Order]] = None, limit: Optional[int] = None, connection: Optional[DatabaseConnection] = None) -> Sequence[TokenAttribute]:
```
Comment (krishan711) [inline api/notd/store/retriever.py:225]: move function below

File: api/notd/token_attributes_processor.py [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+
+from notd.model import RetrievedTokenAttribute
+from notd.model import TokenMetadata
+from notd.store.retriever import Retriever
+from notd.store.schema import TokenMetadatasTable
+
+
+class TokenAttributeProcessor:
+
+    def __init__(self,retriever: Retriever) -> None:
```
Comment (krishan711) [inline api/notd/token_attributes_processor.py:16]:
```suggestion
    def __init__(self, retriever: Retriever) -> None:
```

File: api/notd/token_attributes_processor.py [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+        tokenMetadata: List[TokenMetadata] = await self.retriever.list_token_metadatas(
+            fieldFilters=[
+                StringFieldFilter(TokenMetadatasTable.c.registryAddress.key, eq=registryAddress),
+                StringFieldFilter(TokenMetadatasTable.c.tokenId.key, eq=tokenId),
+            ]
+        )
+        token = tokenMetadata[0]
+        tokenAttributes = []
+        for attribute in token.attributes:
+            tokenAttributes += [RetrievedTokenAttribute(registryAddress=token.registryAddress, tokenId=token.tokenId, attributeName=attribute.get("trait_type"), attributeValue=attribute.get("value"))]
```
Comment (krishan711) [inline api/notd/token_attributes_processor.py:29]: this is wrong, name and value cannot be nullable

File: dbschema/1.tables.sql [hunk 24]
```diff
@@ -209,13 +209,27 @@ CREATE INDEX tbl_latest_updates_date ON tbl_latest_updates (date);
 CREATE INDEX tbl_latest_updates_key ON tbl_latest_updates (key);
 CREATE INDEX tbl_latest_updates_name ON tbl_latest_updates (name);
 
-CREATE TABLE tbl_latest_token_listings (
+CREATE TABLE tbl_token_attributes (
```
Comment (krishan711) [inline dbschema/1.tables.sql:212]: move down

File: dbschema/1.tables.sql [hunk 25]
```diff
@@ -209,13 +209,27 @@ CREATE INDEX tbl_latest_updates_date ON tbl_latest_updates (date);
...
     id BIGSERIAL PRIMARY KEY,
     created_date TIMESTAMP WITHOUT TIME ZONE NOT NULL,
     updated_date TIMESTAMP WITHOUT TIME ZONE NOT NULL,
     registry_address TEXT NOT NULL,
     token_id TEXT NOT NULL,
+    attribute_name TEXT,
+    attribute_value TEXT
```
Comment (krishan711) [inline dbschema/1.tables.sql:219]: nullable

File: api/notd/token_manager.py [hunk 26]
```diff
@@ -378,6 +384,9 @@ async def reprocess_owner_token_ownerships(self, ownerAddress: str) -> None:
     async def update_activity_for_all_collections_deferred(self) -> None:
         await self.tokenQueue.send_message(message=UpdateActivityForAllCollectionsMessageContent().to_message())
 
+    async def update_token_attributes_for_all_collections_deferred(self) -> None:
```
Comment (krishan711) [inline api/notd/token_manager.py:387]: move this to where the other functions are

File: api/notd/token_manager.py [hunk 27]
```diff
@@ -410,6 +419,28 @@ async def update_activity_for_all_collections(self) -> None:
         await self.tokenQueue.send_messages(messages=messages)
         await self.saver.update_latest_update(latestUpdateId=latestUpdate.latestUpdateId, date=startDate)
 
+    async def update_token_attributes_for_all_collection(self) -> None:
+        for _, address in enumerate(GALLERY_COLLECTIONS):
```
Comment (krishan711) [inline api/notd/token_manager.py:423]:
```suggestion
        for address in GALLERY_COLLECTIONS:
```

File: api/notd/token_manager.py [hunk 28]
```diff
@@ -410,6 +419,28 @@ async def update_activity_for_all_collections(self) -> None:
...
+            await self.update_token_attributes_for_all_collection_deferred(address=address)
+
+    async def update_token_attributes_for_all_collection_deferred(self, address: str) -> None:
```
Comment (krishan711) [inline api/notd/token_manager.py:426]: this name is bad

File: api/notd/token_manager.py [hunk 29]
```diff
@@ -410,6 +419,28 @@ async def update_activity_for_all_collections(self) -> None:
         await self.tokenQueue.send_messages(messages=messages)
         await self.saver.update_latest_update(latestUpdateId=latestUpdate.latestUpdateId, date=startDate)
 
+    async def update_token_attributes_for_all_collection(self) -> None:
```
Comment (krishan711) [inline api/notd/token_manager.py:422]: name is bad

File: api/notd/token_manager.py (hunk 28 above)
Comment (krishan711) [reply api/notd/token_manager.py:426]: this function looks super wrong

File: api/notd/token_manager.py [hunk 30]
```diff
@@ -435,6 +466,40 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
                 else:
                     await self.saver.create_collection_hourly_activity(connection=connection, address=retrievedCollectionActivity.address, date=retrievedCollectionActivity.date, transferCount=retrievedCollectionActivity.transferCount, saleCount=retrievedCollectionActivity.saleCount, totalValue=retrievedCollectionActivity.totalValue, minimumValue=retrievedCollectionActivity.minimumValue, maximumValue=retrievedCollectionActivity.maximumValue, averageValue=retrievedCollectionActivity.averageValue,)
 
+    async def update_token_attributes_for_collection_deferred(self, registryAddress: str, tokenId: str) -> None:
```
Comment (krishan711) [inline api/notd/token_manager.py:469]: move all the code together

Comment (krishan711) [reply api/notd/token_manager.py:469]: name is wrong

File: api/notd/token_manager.py [hunk 31]
```diff
@@ -435,6 +466,40 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+        await self.tokenQueue.send_messages(messages=UpdateTokenAttributesForCollectionMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message())
+
+    async def update_token_attributes_for_collection(self, registryAddress: str, tokenId: str) -> None:
```
Comment (krishan711) [inline api/notd/token_manager.py:472]: name is wrong

File: api/notd/token_manager.py [hunk 32]
```diff
@@ -435,6 +466,40 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+        async with self.saver.create_transaction() as connection:
+            retrievedTokenAttributes: List[RetrievedTokenAttribute] = await self.tokenAttributeProcessor.get_token_attributes(registryAddress=registryAddress, tokenId=tokenId)
```
Comment (krishan711) [inline api/notd/token_manager.py:474]: either pass the connection in or move this out

File: api/notd/notd_message_processor.py [hunk 33]
```diff
@@ -69,6 +71,17 @@ async def process_message(self, message: SqsMessage) -> None:
             messageContent = UpdateActivityForCollectionMessageContent.parse_obj(message.content)
             await self.notdManager.update_activity_for_collection(address=messageContent.address, startDate=messageContent.startDate)
             return
+        if message.command == UpdateTokenAttributesForAllCollectionsMessageContent.get_command():
+            if message.postDate is None or message.postDate < date_util.datetime_from_now(seconds=-(60 * 10)):
```
Comment (krishan711) [inline api/notd/notd_message_processor.py:75]: this is gonna run every hour, please update this accordingly

Comment (krishan711) [reply api/notd/notd_message_processor.py:75]: could you do the same for `UpdateListingsForAllCollections` below as well please

File: api/notd/store/schema.py (hunk 14 above)
Comment (krishan711) [reply api/notd/store/schema.py:164]: what ahppening here??

File: api/notd/store/saver.py [hunk 34]
```diff
@@ -477,6 +479,47 @@ async def update_latest_update(self, latestUpdateId: int, key: Optional[str] = N
         query = LatestUpdatesTable.update(LatestUpdatesTable.c.latestUpdateId == latestUpdateId).values(values)
         await self._execute(query=query, connection=connection)
 
+    async def create_token_attribute(self, registryAddress: str, tokenId: str, name: Optional[str], value: Optional[str], connection: Optional[DatabaseConnection] = None) -> TokenAttribute:
```
Comment (krishan711) [inline api/notd/store/saver.py:482]: name should not be optional here

File: api/notd/store/saver.py [hunk 35]
```diff
@@ -477,6 +479,47 @@ async def update_latest_update(self, latestUpdateId: int, key: Optional[str] = N
...
+            tokenId=tokenId,
+            name=name,
+            value=value,
+        )
+
+    async def update_token_attribute(self, tokenAttributeId: int, name: Optional[str] = _EMPTY_STRING, value: Optional[str] = _EMPTY_STRING, connection: Optional[DatabaseConnection] = None) -> None:
+        values = {}
+        if name != _EMPTY_STRING:
+            values[TokenAttributesTable.c.name.key] = name
+        if value is not None:
```
Comment (krishan711) [inline api/notd/store/saver.py:510]: this is wrong - nullable fields should use the _EMPTY_STRING thing and non-nullable should not.

File: api/notd/token_attributes_processor.py [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+from notd.store.schema import TokenMetadatasTable
+
+
+class TokenAttributeProcessor:
+
+    def __init__(self, retriever: Retriever) -> None:
+        self.retriever = retriever
+
+    async def get_token_attributes(self, registryAddress: str, tokenId: str) -> List[TokenAttribute]:
+        tokenMetadata: List[TokenMetadata] = await self.retriever.list_token_metadatas(
```
Comment (krishan711) [inline api/notd/token_attributes_processor.py:19]: why not use the direct get fucntion?

File: api/notd/token_attributes_processor.py [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+            fieldFilters=[
+                StringFieldFilter(TokenMetadatasTable.c.registryAddress.key, eq=registryAddress),
+                StringFieldFilter(TokenMetadatasTable.c.tokenId.key, eq=tokenId),
+            ]
+        )
+        token = tokenMetadata[0]
+        tokenAttributes = []
+        for attribute in token.attributes:
+            name, value = list(attribute.values())[0], list(attribute.values())[1]
```
Comment (krishan711) [inline api/notd/token_attributes_processor.py:28]: this looks very wierd, can you explain whats happening here pls, im sure this code could be clearer

File: api/notd/token_manager.py [hunk 38]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
                 else:
                     await self.saver.create_collection_hourly_activity(connection=connection, address=retrievedCollectionActivity.address, date=retrievedCollectionActivity.date, transferCount=retrievedCollectionActivity.transferCount, saleCount=retrievedCollectionActivity.saleCount, totalValue=retrievedCollectionActivity.totalValue, minimumValue=retrievedCollectionActivity.minimumValue, maximumValue=retrievedCollectionActivity.maximumValue, averageValue=retrievedCollectionActivity.averageValue,)
 
+
```
Comment (krishan711) [inline api/notd/token_manager.py:444]: rm one empty line

File: api/notd/token_manager.py [hunk 39]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+    async def update_token_attributes_for_all_collections_deferred(self) -> None:
+        await self.tokenQueue.send_message(message=UpdateTokenAttributesForAllCollectionsMessageContent().to_message())
+
+    async def update_token_attributes_for_all_collections(self) -> None:
+        for _, address in enumerate(GALLERY_COLLECTIONS):
```
Comment (krishan711) [inline api/notd/token_manager.py:449]:
```suggestion
        for address in GALLERY_COLLECTIONS:
```

File: api/notd/token_manager.py [hunk 40]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+        messages = [UpdateCollectionTokenAttributesMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message() for (registryAddress, tokenId) in updatedTokenMetadatas]
+        await self.tokenQueue.send_messages(messages=messages)
+        await self.saver.update_latest_update(latestUpdateId=latestUpdate.latestUpdateId, date=startDate)
+
+    async def update_collection_token_attributes_deferred(self, registryAddress: str, tokenId: str) -> None:
+        await self.tokenQueue.send_messages(messages=UpdateCollectionTokenAttributesMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message())
+
+    async def update_collection_token_attributes(self, registryAddress: str, tokenId: str) -> None:
+        retrievedTokenAttributes: List[RetrievedTokenAttribute] = await self.tokenAttributeProcessor.get_token_attributes(registryAddress=registryAddress, tokenId=tokenId)
+        tokenAttributes = await self.retriever.list_token_attributes(
```
Comment (krishan711) [inline api/notd/token_manager.py:475]: this should be inside and use the connection (only this line not the one above

File: api/notd/token_manager.py [hunk 41]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+        logging.info(f'Scheduling processing for {len(updatedTokenMetadatas)} tokens in collection {address}')
+        messages = [UpdateCollectionTokenAttributesMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message() for (registryAddress, tokenId) in updatedTokenMetadatas]
+        await self.tokenQueue.send_messages(messages=messages)
+        await self.saver.update_latest_update(latestUpdateId=latestUpdate.latestUpdateId, date=startDate)
+
+    async def update_collection_token_attributes_deferred(self, registryAddress: str, tokenId: str) -> None:
+        await self.tokenQueue.send_messages(messages=UpdateCollectionTokenAttributesMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message())
+
+    async def update_collection_token_attributes(self, registryAddress: str, tokenId: str) -> None:
+        retrievedTokenAttributes: List[RetrievedTokenAttribute] = await self.tokenAttributeProcessor.get_token_attributes(registryAddress=registryAddress, tokenId=tokenId)
```
Comment (krishan711) [inline api/notd/token_manager.py:474]: why do you need to explicity typing on this line? it should be coming from the typing of the function

File: api/notd/token_manager.py [hunk 42]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+                StringFieldFilter(fieldName=TokenAttributesTable.c.tokenId.key, eq=tokenId),
+            ]
+        )
+        async with self.saver.create_transaction() as connection:
+            if tokenAttributes:
+                existingTokenAttributesTuples = {(existingTokenAttribute.registryAddress, existingTokenAttribute.tokenId, existingTokenAttribute.name): existingTokenAttribute.tokenAttributeId for existingTokenAttribute in tokenAttributes }
+                tokenAttributeIdsToDelete = (list(existingTokenAttributesTuples.values()))
+                logging.info(f'Deleting  {len(tokenAttributeIdsToDelete)} attributes for registryAddress: {registryAddress}, tokenId: {tokenId}')
+                await self.saver.delete_token_attributes(connection=connection, tokenAttributeIds=tokenAttributeIdsToDelete)
+                #NOTE Didn't use comprehension cause connection kept closing
```
Comment (krishan711) [inline api/notd/token_manager.py:487]: formoat these properly: `# NOTE(username): message` and `# TODO(username): message`

File: api/notd/token_manager.py [hunk 43]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+                #TODO Account for only updated collection token attribute
+                for retrievedTokenAttribute in retrievedTokenAttributes:
```
Comment (krishan711) [inline api/notd/token_manager.py:489]: this is all repeated in the else block, just remove it

File: api/notd/token_manager.py [hunk 44]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+    async def update_collection_token_attributes(self, registryAddress: str, tokenId: str) -> None:
+        retrievedTokenAttributes: List[RetrievedTokenAttribute] = await self.tokenAttributeProcessor.get_token_attributes(registryAddress=registryAddress, tokenId=tokenId)
+        tokenAttributes = await self.retriever.list_token_attributes(
+            fieldFilters=[
+                StringFieldFilter(fieldName=TokenAttributesTable.c.registryAddress.key, eq=registryAddress),
+                StringFieldFilter(fieldName=TokenAttributesTable.c.tokenId.key, eq=tokenId),
+            ]
+        )
+        async with self.saver.create_transaction() as connection:
+            if tokenAttributes:
```
Comment (krishan711) [inline api/notd/token_manager.py:482]: why is this needed? it wuold just be an empty list and `delete_token_attributes` should just do nothin if its and empty list

File: api/notd/token_manager.py (hunk 43 above)
Comment (Femi-Ogunkola) [reply api/notd/token_manager.py:489]: I thought we delete then add back, if I remove it won't it just delete

File: api/notd/token_manager.py [hunk 45]
```diff
@@ -435,6 +441,72 @@ async def update_activity_for_collection(self, address: str, startDate: datetime
...
+                await self.saver.delete_token_attributes(connection=connection, tokenAttributeIds=tokenAttributeIdsToDelete)
+                #NOTE Didn't use comprehension cause connection kept closing
+                #TODO Account for only updated collection token attribute
+                for retrievedTokenAttribute in retrievedTokenAttributes:
+                    logging.info(f'Saving  {len(tokenAttributeIdsToDelete)} attributes for registryAddress: {registryAddress}, tokenId: {tokenId}')
+                    await self.saver.create_token_attribute(connection=connection, registryAddress=registryAddress, tokenId=tokenId, name=retrievedTokenAttribute.name, value=retrievedTokenAttribute.value)
+            else:
+                for retrievedTokenAttribute in retrievedTokenAttributes:
+                    logging.info(f'Creating  {len(tokenAttributeIdsToDelete)} attributes for registryAddress: {registryAddress}, tokenId: {tokenId}')
+                    await self.saver.create_token_attribute(connection=connection, registryAddress=registryAddress, tokenId=tokenId, name=retrievedTokenAttribute.name, value=retrievedTokenAttribute.value)
```
Comment (krishan711) [inline api/notd/token_manager.py:495]:
```suggestion
            existingTokenAttributesTuples = {(existingTokenAttribute.registryAddress, existingTokenAttribute.tokenId, existingTokenAttribute.name): existingTokenAttribute.tokenAttributeId for existingTokenAttribute in tokenAttributes }
            tokenAttributeIdsToDelete = (list(existingTokenAttributesTuples.values()))
            logging.info(f'Deleting  {len(tokenAttributeIdsToDelete)} attributes for registryAddress: {registryAddress}, tokenId: {tokenId}')
            await self.saver.delete_token_attributes(connection=connection, tokenAttributeIds=tokenAttributeIdsToDelete)
            #NOTE Didn't use comprehension cause connection kept closing
            #TODO Account for only updated collection token attribute
            for retrievedTokenAttribute in retrievedTokenAttributes:
                logging.info(f'Saving  {len(tokenAttributeIdsToDelete)} attributes for registryAddress: {registryAddress}, tokenId: {tokenId}')
                await self.saver.create_token_attribute(connection=connection, registryAddress=registryAddress, tokenId=tokenId, name=retrievedTokenAttribute.name, value=retrievedTokenAttribute.value)
```

**Review by krishan711 — APPROVED** [review]
